### ISSUE TITLE
Loop action

### ISSUE THREAD

**Issue by dylburger** [PipedreamHQ/pipedream#193]
I'd like to run a set of steps or actions within a loop construct. For example, I might receive an array of 100 events in my HTTP payload, and I need to run an action on each. I can do this with code, but I have no way to loop through the elements of this array using Pipedream's actions. I'd like a "loop" action that can handle this use case.

**Comment by tinjaw** [comment]
As a temporary workaround, can you please add the ability to duplicate an existing step in a workflow? So, if I want to do something ten times, I can create the action, duplicate it nine times, and then just modify the parameters in the duplicate actions.

**Comment by pauliusuza** [comment]
This functionality is a must! If a step receives an array, it should simply run the same step in the sequence for X times.

**Comment by alexanderahlsen** [comment]
What is the status on this? It's a killer for when moving from Make.

**Comment by dylburger** [comment]
@alexanderahlsen This is on the backlog for 2022! We're working on a GitHub integration now, and plan to work on control flow (loops, conditionals, etc.) soon after!

**Comment by gustavorps** [comment]
This is what is preventing me from migrating from Integromat

**Comment by ctrlaltdylan** [comment]
Here's a temporary workaround by utilizing two different workflows.

The first workflow is the _fetching_ workflow. It grabs this array of records from Airtable, Google Sheets, Quickbooks, or some database or API.

The second workflow is the _processing_ workflow. It's responsible for processing a _single_ record at a time, so that way you can use all of the pre-built actions Pipedream has to offer.


## How to create a processing workflow

First, create a new workflow with an HTTP trigger:


![CleanShot 2022-08-03 at 09 06 48](https://user-images.githubusercontent.com/2694734/182615267-dca3c7cf-5404-4f24-a97d-1ac33223bd15.png)

In this case, the HTTP endpoint for our processing workflow is `https://eo2b9ksk3fy3nrb.m.pipedream.net`.

## How to connect the fetching workflow to the processing workflow

In your fetching workflow after the fetching step, whether that be a trigger or an action add a new Node.js code step and paste in this snippet:

```javascript
import axios from '@pipedream/platform';
// To use previous step data, pass the `steps` object to the run() function
export default defineComponent({
  props: {
    records: {
      type: "array",
      label: "Records to loop",
      description: "The array of records to send to processing workflow",
    },
    workflow_url: {
      type: "string",
      label: "Processing workflow URL",
      description: "The HTTP endpoint to connect to that's processing single individual records from this workflow"
    }
  },
  async run({ steps, $ }) {
    for await(const record of this.records) {
      await axios($, {
        url: this.workflow_url,
        method: 'POST',
        data: record
      })
    }
  },
})

```

Then populate the `Records` prop with the array of records you need processed, and paste in the URL endpoint to your processing workflow. In our example above, the URL would be `https://eo2b9ksk3fy3nrb.m.pipedream.net`.

After clicking *Test* in your fetching workflow, you should see individual HTTP requests hitting your processing workflow with each record individual in the body of the request.

Now you can process each record individually, and leverage the hundreds of pre-built actions in your processing workflow.

**Comment by bacherjo** [comment]
Hey ctrlaltdylan, you made my day! This is an excellent solution to the missing iterator feature in pipedream!
Thank you!
Johannes

**Comment by djpecot** [comment]
I ran into an issue where my API came back as an `object` instead of an `array`. This modified code should handle single values relatively gracefully 🤞 @ctrlaltdylan 
```
import { axios } from '@pipedream/platform';
// To use previous step data, pass the `steps` object to the run() function
export default defineComponent({
  props: {
    records: {
      type: "any",
      label: "Records to loop",
      description: "The array of records to send to processing workflow",
    },
    workflow_url: {
      type: "string",
      label: "Processing workflow URL",
      description: "The HTTP endpoint to connect to that's processing single individual records from this workflow"
    }
  },

  
  

  async run({ steps, $ }) {
    // Break this if a single value (Array | Object)
    if (Array.isArray(this.records)) {
      for await(const record of this.records) {
        await axios($, {
          url: this.workflow_url,
          method: 'POST',
          data: record
        })
      }
    } else {
      await axios($, {
          url: this.workflow_url,
          method: 'POST',
          data: this.records
        })
    }
  },
})
```

**Comment by ctrlaltdylan** [comment]
Hi @djpecot , yup I follow!

Though I'm curious how your data is being passed to the `records` prop is sometimes an array and sometimes a single object.

What's the source of your data if you don't mind me asking?

**Comment by djpecot** [comment]
@ctrlaltdylan It's a webhook that's hitting me with a variable amount of objects (Toggl). It was hitting me with either an array or a single object as I recall, so had to account for it.

**Comment by LucBerge** [comment]
> This functionality is a must! If a step receives an array, it should simply run the same step in the sequence for X times.

I'm looking for this feature as well 😄

**Comment by tqbdev** [comment]
> This functionality is a must! If a step receives an array, it should simply run the same step in the sequence for X times.

Any update for this feature?

**Comment by DFazekas** [comment]
Just created an account on this website an hour ago, and already deleting my account due to the lack of this feature. What a shame.

**Comment by zvictor** [comment]
> Just created an account on this website an hour ago, and already deleting my account due to the lack of this feature. What a shame.

Exactly my case as well 😅.
Pipedream seems to be the best solution, otherwise! ❤️

**Comment by bawantha** [comment]
3 years and not yet implemented

**Comment by bawantha** [comment]
> @alexanderahlsen This is on the backlog for 2022! We're working on a GitHub integration now, and plan to work on control flow (loops, conditionals, etc.) soon after!

like people care about GitHub integration?

**Comment by blendedmarket** [comment]
I think this has to be the deal breaker for most users once they realize this limitation. Perhaps its because making an extra call to a different workflow allows them to bill you separately for each iteration? Correct me if I am wrong , but each iteration would be an additional [invocation](https://pipedream.com/docs/pricing/#invocations)? 
It's a shame, because I was really enjoying the building process only to find out I can't afford to use it.

**Comment by bacherjo** [comment]
Yes I can confirm that calling a different workflow counts as a separate invocation. I have to process orders with 20 items, so instead of having one invocation i have 21. 
Still I think we need this feature and need the ability to handle this within one invocation. It gets confusing and hard to monitor with a lot of single workflows.

**Comment by Landsil** [comment]
eh, back to Zapier I guess :( 
I was really looking towards switching away from them.

I can code loop myself but if have to bother with that I may as well code the rest and self host it instead of pushing data via external services...

**Comment by eigenstil** [comment]
Just a quick note for other people stumbling accross this issue: 

following @ctrlaltdylan's step-by-step [explanation above](https://github.com/PipedreamHQ/pipedream/issues/193#issuecomment-1203934676), I was able to solve it for now. Though, this is somewhat tedious ;-)

**Comment by acreconsulting** [comment]
Just wanted to add my +1!

**Comment by evelant** [comment]
A no-code way to do simple iteration over array elements would be really nice! There are lots of cases where you can end up with an array of data but the API you're working with won't accept array/bulk input so you need to send a request for each one.

**Comment by LucBerge** [comment]
I guess there is nothing technical preventing them to implement the loop system. What could limit them to do it is the business model: Currently, user/organization pay per event. The problem is that if a step retrieves 100 elements to process in a loop, it is still counted as one event. Since pipedream is probably using a pay as you go server (like AWS), they need to think about a new business model before going further. This is probably why this issue is in standby since 4 years...

**Comment by dylburger** [comment]
Hi y'all, just want to let you know we read all this feedback and very much appreciate the use cases! This is on our roadmap for this year, after we ship our Git integration, folders, and other big features we're working on right now. So we plan to address this soon.

**Comment by LucBerge** [comment]
> folders

Hello, great news 🎉 Will the tiers change ?

**Comment by dylburger** [comment]
We haven't decided exactly how this will impact the invocations / credits model, but we'll comment here / share in future announcements on all the details.

**Comment by sammymlangeni** [comment]
There was one stage where I was looking to get the payload from Woocommerce. I was more interested in getting the licenses keys to uploaded to Dropbox folder. My use case may not be as complicated. The payload created 40 lines of code of which will never be the real scenario for my situation. 

I am sure this is only applicable to me.
![](https://pipedream.com/community/uploads/default/original/2X/a/acbfa82c58b6c1e7f1efe4b4d7e84e62679ec7a9.jpeg)

**Comment by DhavalW** [comment]
Seems like this is still a pipedream eh.

+1 its unfair if each looped item counts as a new invocation and gets billed.

**Comment by LucBerge** [comment]
> +1 its unfair if each looped item counts as a new invocation and gets billed. I love pipedream so far, but If this happens, I'm moving to autocode or something else

Seems like invocation system is beeing replaced by a credit system.

1 credit = 30 sec (256 Mo)
100 credits/day (for free tier) = 50min/day

This is still better than other intergration platforms

**Comment by DhavalW** [comment]
Yes. Definitely. One of the key things I like about pipedream for sure.

But AFAIK it's not fully replaced.

Each invocation still counts as min.1 credit, regardless of how long you
use it.

So if you have 10 invocations of an api that last 2s each - it bills 10
credits, even though total runtime across calls is < 1 credit.

**Comment by NebularNerd** [comment]
Just came across a workflow where I need this, I can likely create a bodge of some description with a dual workflow but this would seem to be a must have helper function.

**Comment by ExpertLearner7** [comment]
Hi, we are trying to move from Zapier to Pipedream but we use the loop action in some Zaps. Is there a no code way to iterate through an array/list? to do some actions repeatedly over each array/list element?

**Comment by DhavalW** [comment]
Actually i realised later that you can. Needs to be custom coded through a
node js block.

But it's been a while, so not sure if it's still possible or killed by the
recent platform updates.

**Comment by RichardBaoLei** [comment]
Just quickly looked through each posts. the no-code loops step is really a deal breaker. the current nodeJS solution is not elegant and user-friendly for most people. is there any definitive date for the development of it? @vunguyenhung @dylburger @ctrlaltdylan

**Comment by thomasbiddle** [comment]
How's this coming along? I'm considering migrating from Make to Pipedream due to some frustrations; but this is a dealbreaker.

**Comment by nadyyym** [comment]
Any estimates on this? Want to move all my automations to Pipedream, but I am not technical and would really hate to implement loops with code

**Comment by nadyyym** [comment]
> loop option has many potential variations

> a lot of flexibility and was not too hard to do

Fair enough. Guess I'll refresh my Python skills.

**Comment by motiejunas** [comment]
> @alexanderahlsen This is on the backlog for 2022! We're working on a GitHub integration now, and plan to work on control flow (loops, conditionals, etc.) soon after!

Late 2023 already, so I wonder what is the progress?

**Comment by panda-inc-1** [comment]
Pipedream will become our EXCLUSIVE low code solution (we currently pay for Make, Zapier, n8n) if it incorporates advanced logic like this!

**Comment by keladeine** [comment]
I'm just trying to transfer tags of an item. It's a really big flaw if I have to run a workflow 4 times just because one record has multiple tags.

**Comment by ctrlaltdylan** [comment]
Thanks for the candid feedback everyone, and we really appreciate your patience for this highly requested feature.

We know that Control Flow unlocks the ability for your workflows from a linear to a complex branching and looping experience and it would be a massive productivity gain.

This feature is the highest priority for our team, and we're actively working to release a beta version in the near feature. 

If you're interested in gaining early access to this major feature that would allow branching and looping, join by filling out this short beta interest form so we can contact you as soon as it's available:

https://forms.pipedream.com/control-flow-beta

**Comment by dylburger** [comment]
@jscottcronin Nothing we love more than to wake up and read comments like this!

A GitHub issue is not a promise. This was among the first requests we received when we launched the platform, so we created this issue. In the meantime, we’ve shipped hundreds of meaningful platform features and thousands of integrations. There will always be features like this that take time to prioritize and ship. Every platform deals with this.

I empathize so much with the desire for flow control and completely feel the pain. Try to articulate that feedback without calling Pipedream “a joke” and trashing our technical team. Choose kindness! You’ll get so much more mileage with it.

**Comment by mroy-seedbox** [comment]
@jscottcronin: Did you know that Zapier existed for **_seven years_** without the ability to do even the most basic branching? Let alone looping, which was released years later. They were officially founded in 2011, and only added branches ("paths") in 2018.

Pipedream is still a pretty young company, but give them some time and they'll knock it out of the park -- in half the time it took for Zapier to develop these features, and with many much more awesome ones.

On our side, we're migrating away from Zapier. And we're not looking back.

But I do understand the impatience for this feature.

**Comment by andreaskweber** [comment]
So, any update on this? :-)

**Comment by rmcsharry** [comment]
@dylburger I loved your response above about "waking up" to such negative comments.

I am with you 100% but you did say in March 2023 "So we plan to address this soon."

I'd wager that "soon" kinda implies < 12 months to most people. 

An update would be greatly appreciated. How soon? 😉 

Thanks!
PS Pipedream is still awesome!

**Comment by dylburger** [comment]
@rmcsharry definitely taking longer than we’d like, but we just shipped the first version of conditional branching and are actively working on loops! If you’re interested in testing, we typically release these for early feedback on our Slack community (https://pipedream.com/support). Take a look at the #announcements channel there for details.

**Comment by rmcsharry** [comment]
@dylburger Thanks for the considerate and timely reply! I've not had time to test this, hopefully soon. 🤞

**Comment by matthewt** [comment]
N8n has a nice way of working where if a step receives an array it automatically applies that step to each element in the array. If it receives an object it runs the step once. Its really good to work that way so that your flows don’t get disrupted when there is a change in multiplicity, and there is less “noise” in the workflow design.

I agree this is a critical requirement though I can’t believe such a nice and well build product doesn’t have a good solution. Having to drop into code for the most basic need isn’t nice.  (posted here 7 months ago.. .https://pipedream.com/community/t/looping-through-results/3217/3)

**Comment by dimaip** [comment]
> N8n has a nice way of working where if a step receives an array it automatically applies that step to each element in the array. If it receives an object it runs the step once. Its really good to work that way so that your flows don’t get disrupted when there is a change in multiplicity, and there is less “noise” in the workflow design.
> 
> I agree this is a critical requirement though I can’t believe such a nice and well build product doesn’t have a good solution. Having to drop into code for the most basic need isn’t nice. (posted here 7 months ago.. .https://pipedream.com/community/t/looping-through-results/3217/3)

Came here to say exactly this: n8n does this nicely.
